rdt/hyper_transformer.py

```diff
@@ -17,136 +10,161 @@ class HyperTransformer:
+        if output in self.field_transformers:
+            if self._field_in_data(output, data):
+                next_transformer = self.field_transformers[output]
 
-        Returns:
-            dict:
-                Mapping of column names and transformer instances.
+        elif next_transformers is not None:
+            next_transformer = next_transformers[output]
 
-        Raises:
-            ValueError:
-                if a ``dtype`` is not supported by the `HyperTransformer``.
-        """
-        transformers = {}
-        if self.dtypes:
-            dtypes = self.dtypes
         else:
-            dtypes = [
-                data[column].dropna().infer_objects().dtype
-                for column in data.columns
-            ]
-
-        for name, dtype in zip(data.columns, dtypes):
-            try:
-                kind = np.dtype(dtype).kind
-            except TypeError:
-                # probably category
-                kind = 'O'
-
-            transformer_template = self.dtype_transformers.get(kind)
-            if not transformer_template:
-                raise ValueError('Unsupported dtype: {}'.format(dtype))
-
-            if isinstance(transformer_template, str):
-                transformer_template = self._TRANSFORMER_TEMPLATES[transformer_template]
-
-            if not isinstance(transformer_template, type):
-                transformer = deepcopy(transformer_template)
-            else:
-                transformer = transformer_template()
+            next_transformer = get_default_transformer(output_type)
+
+        return next_transformer
 
-            transformers[name] = transformer
+    def _fit_field_transformer(self, data, field, transformer):
+        """Fit a transformer to its corresponding field.
 
-        return transformers
+        If the transformer outputs fields that aren't ML ready, then this method
+        recursively fits transformers to their outputs until they are. This method
+        keeps track of which fields are temporarily created by transformers as well
+        as which fields will be part of the final output from ``transform``.
+
+        Args:
+            data (pandas.DataFrame):
+                Data to fit the transformer to.
+            field (str or tuple):
+                Name of column or tuple of columns in data that will be transformed
+                by the transformer.
+            transformer (Transformer):
+                Instance of transformer class that will fit the data.
+        """
+        transformer = load_transformer(transformer)
+        transformer.fit(data, field)
+        self._transformers_sequence.append(transformer)
+        output_types = transformer.get_output_types()
+        next_transformers = transformer.get_next_transformers()
+        for (output, output_type) in output_types.items():
+            output_field = self._multi_column_fields.get(output, output)
+            if (output_field not in self.field_transformers
```

Comment:
I added this check first because I realized that in every other case we still want to run `transformer.transform`, but there are two cases where the `next_transformer` is `None`: 
1. If the column is part of a multi-column field and the other fields aren't ready
2. If we are done with the output
In the first case, we still need to transform the data but in the second case we don't want to and with the method we discussed it was hard to distinguish

Reply:
Interesting. As we discussed, I think this approach is right and the only comment here would be to maybe try to make it a bit more readable by renaming `next_transformer = get_next_transformer` to `next_transformer, transform = get_next_steps`, so all the `if/else` conditions are inside the same method.

Reply:
I am now realizing there is a problem with this approach. We are looping through the output types for the transformer, and the `data = transformer.transform()` call is in the loop, meaning potentially the same transformer might attempt to transform twice. 

I think either we have to just call `transformer.transform()` outside the loop, or check in the beginning if a transform is needed which is if any of the `output_types` are not final or if any of the `output_columns` are in `field_transformers`. Thoughts?

Reply:
Or a third approach which is to check if the transformation was already done or not. That's probably what I'll go with

Reply:
I decided to just return a `next_transformer` whenever there is one. That way it is only null if the output field is final. Then the caller can decide whether to use the `next_transformer` or not based on if all the columns necessary are present in the data